scripts/upload_samples.py

```diff
@@ -0,0 +1,95 @@
+""" Updates the Sample Metadata Database to register new samples """
+import csv
+import io
+import os
+import json
+import subprocess
+from typing import List, Tuple
+import click
+from google.cloud import storage
+
+from sample_metadata.api.sample_api import SampleApi
+
+
+def get_csv(bucket_name, prefix) -> Tuple[csv.DictReader, str]:
+    """Pull relevant data from csv file in a given bucket
+    ==========
+    Only one .csv file exists in the upload bucket and previous batch directory.
+    """
+    client = storage.Client()
+    gcs_bucket = client.get_bucket(bucket_name)
+
+    all_blobs = list(client.list_blobs(bucket_name, prefix=prefix))
+    csv_path = next(filter(lambda blob: blob.name.endswith('.csv'), all_blobs)).name
+
+    csv_as_text = gcs_bucket.get_blob(csv_path).download_as_text()
+    csv_reader = csv.DictReader(io.StringIO(csv_as_text))
+
+    return csv_reader, csv_path
+
+
+def update_samples(csv_dict_reader):
+    """Update the sample status
+    When the initial sample list is uploaded (before all sequencing has been uploaded) the status for
+    each sample should reflect this i.e. 'New' or 'Registered' or 'Waiting' or something.
+    Then, when the sample is uploaded this sample status should change to 'Active'.
+    This function will decide what samples should have this status change to upload."""
+
+    # Pull a list containing the sample ID's from the latest .mt.
+    sapi = SampleApi()
+    mt_samples = sapi.get_samples_from_latest_analysis('mt')  # TODO: Implement.
+
+    # The samples list from the csv file is cumulative.
+
+    # Pull the data from the CSV iterable.
+
+    samples: List[str] = []
+    sample_metadata = []
+
+    for sample_dict in csv_dict_reader:
+        samples.append(sample_dict['sample.sample_name'])
+        sample_metadata.append(sample_dict)
+
+    # deleted_samples = mt_samples - samples
+
+    # Determine the samples in the latest upload.
+    latest_upload = samples - mt_samples
+
+    for sample in latest_upload:
+        # pylint: disable=W0640
+        current_metadata = filter(
```

Comment:
Instead of filtering the array of `sample_metadata` to find the ID for each time, consider using a map, to first key by the sample_name, and then you can pull it out in one hit, eg:

```python
sample_meta_map = {d['sample.sample_name']: d for d in sample_metadata}
for sample in latest_upload:
    metadata = sample_meta_map[sample]
    # ...
```

Reply:
That's much nicer! Thank you :)